src/python/WMCore/MicroService/Unified/MSTransferor.py

```diff
@@ -326,3 +360,105 @@ def createTransferDoc(self, reqName, transferRecords):
+            elif line.startswith('#'):
+                continue
+
+            if not doRead:
+                continue
+
+            _, quota, _, _, pnn = line.split()
+
+            if pnn.endswith("_MSS") or pnn.endswith("_Export"):
+                continue
+            # convert from TB to bytes
+            self.nodesUsage.setdefault(pnn, {})
+            self.nodesUsage[pnn] = dict(quota=int(quota) * (1024 ** 4),
+                                        bytes_limit=0,
+                                        bytes=0,
+                                        bytes_remaining=0,
+                                        quota_avail=0)
+
+
+    def _getStorageUsage(self):
+        """
+        Fetch the storage usage from either Rucio or PhEDEx, which will then
+        be used as part of the data placement mechanism.
+        Also calculate the available quota - given the configurable quota
+        fraction - and mark RSEs with less than 1TB available as NOT usable.
+
+        Keys definition is:
+         * quota: the PhEDEx group quota provided by Detox
+         * bytes_limit: either the PhEDEx quota or the account quota from Rucio
+         * bytes: data volume placed by Rucio (or subscribed in PhEDEx)
+         * bytes_remaining: storage available for our account/group
+         * quota_avail: space left (in bytes) that we can use for data placement
+        :return: update our cache in place with up-to-date values, in the format of:
+            {"pnn_name": {"bytes_limit": total space for the account/group,
+                          "bytes": amount of bytes currently used/archived,
+                          "bytes_remaining": space remaining for the acct/group}
+        """
+        if hasattr(self, "rucio"):
+            for item in self.rucio.getAccountUsage(self.msConfig['rucioAccount']):
+                self.nodesUsage[item['rse']].update({'bytes_limit': item['bytes_limit'],
+                                                     'bytes': item['bytes'],
+                                                     'bytes_remaining': item['bytes_remaining']})
+        else:
+            # for PhEDEx, we have also to remap the key's to keep in sync with Rucio
+            res = self.phedex.getGroupUsage(group=self.msConfig['group'])
+            for item in res['phedex']['node']:
+                if item['name'] not in self.nodesUsage:
+                    continue
+                quota = self.nodesUsage[item['name']]['quota']
+                self.nodesUsage[item['name']].update({'bytes_limit': quota,
+                                                      'bytes': item['group'][0]['dest_bytes'],
+                                                      'bytes_remaining': quota - item['group'][0]['dest_bytes']})
+
+        self.logger.info("Summary of the initial quotas:")
+        for site, info in self.nodesUsage.items():
+            self.logger.info("  %s : %s", site, info)
+
+    def _evaluateQuotaExceeded(self):
```

Comment:
This method is not thread safe. Please pass to it nodeUsage dict and return fullRSes